### PR TITLE
Moving non-production reusable code to test.driver package #6886

### LINKED ISSUE #6886: Moving non-production reusable code to test.driver package
There are methods in utility classes of `common.util` package which are only used by test classes and not in production code. These methods can be moved to appropriate/new classes in `test.driver` package instead to avoid mixing production and non-production code. 
Helper classes and methods in `teammates.client.scripts.util` can also be moved to test.driver package to put non-production reusable code all in a single package.

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #6886

### CODE REVIEW

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 1]
```diff
@@ -728,7 +696,7 @@ public String getValidityInfoForNonNullField(String fieldName, Object value) {
         return value == null ? NON_NULL_FIELD_ERROR_MESSAGE.replace("${fieldName}", fieldName) : "";
     }
 
-    private boolean isUntrimmed(String value) {
+    public static boolean isUntrimmed(String value) {
```
Comment (whipermr5) [inline src/main/java/teammates/common/util/FieldValidator.java]: This needs documentation since it's now public.

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 2]
```diff
@@ -751,7 +719,7 @@ public static boolean isValidEmailAddress(String email) {
         return uniqueElements.size() == elements.size();
     }
 
-    private static String getPopulatedErrorMessage(
+    public static String getPopulatedErrorMessage(
```
Comment (whipermr5) [inline src/main/java/teammates/common/util/FieldValidator.java]: Same here.

File: src/test/java/teammates/test/cases/util/FieldValidatorTest.java [hunk 3]
```diff
@@ -37,38 +38,38 @@ public void testGetValidityInfoForSizeCappedNonEmptyString() {
...
+                        TestStringHelper.generateStringOfLength(maxLength)));
 
-        String tooLongName = StringHelper.generateStringOfLength(maxLength + 1);
+        String tooLongName = TestStringHelper.generateStringOfLength(maxLength + 1);
         assertEquals("invalid: too long",
                      "\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\" is not acceptable to TEAMMATES "
                          + "as a/an my field because it is too long. The value of a/an my field should be no "
                          + "longer than 50 characters. It should not be empty.",
-                     validator.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, maxLength,
+                         TestStringHelper.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, maxLength,
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: The indentation level shouldn't be changed.

File: src/test/java/teammates/test/cases/util/FieldValidatorTest.java [hunk 4]
```diff
@@ -37,38 +38,38 @@ public void testGetValidityInfoForSizeCappedNonEmptyString() {
...
                                                                           tooLongName));
 
         String emptyValue = "";
         assertEquals("invalid: empty",
                      "\"\" is not acceptable to TEAMMATES as a/an my field because it is empty. The value of "
                          + "a/an my field should be no longer than 50 characters. It should not be empty.",
-                     validator.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, maxLength,
+                         TestStringHelper.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, maxLength,
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: Same here.

File: src/test/java/teammates/test/cases/util/FieldValidatorTest.java [hunk 5]
```diff
@@ -37,38 +38,38 @@ public void testGetValidityInfoForSizeCappedNonEmptyString() {
...
                                                                           emptyValue));
 
         String untrimmedValue = " abc ";
         assertEquals("invalid: untrimmed",
                      "The provided my field is not acceptable to TEAMMATES as it contains only whitespace or "
                          + "contains extra spaces at the beginning or at the end of the text.",
-                     validator.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, maxLength,
+                         TestStringHelper.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, maxLength,
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: Same here.

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        String sanitizedValue = SanitizationHelper.sanitizeForHtml(value);
+        if (value.length() > maxLength) {
+            return FieldValidator.getPopulatedErrorMessage(FieldValidator.SIZE_CAPPED_NON_EMPTY_STRING_ERROR_MESSAGE,
+                    sanitizedValue, fieldName, FieldValidator.REASON_TOO_LONG, maxLength);
+        }
+        return "";
+    }
+
+    /**
+     *generate string of given length.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Missing space; also capitalise the first letter.

File: src/test/java/teammates/test/driver/TestTimeHelper.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package teammates.test.driver;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.TimeZone;
+import teammates.common.util.TimeHelper;
+
+/**
+ *A helper class to hold time-related functions (e.g., converting dates to strings etc.).
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestTimeHelper.java]: Missing space.

File: src/test/java/teammates/test/driver/TestTimeHelper.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ *Time zone is assumed as UTC unless specifically mentioned.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestTimeHelper.java]: Same here.

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
+package teammates.test.driver;
+
+import teammates.common.util.Assumption;
+import teammates.common.util.FieldValidator;
+import teammates.common.util.SanitizationHelper;
+import teammates.common.util.StringHelper;
+
+/**
+ *String Helper class for tests.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Missing space. Also, `Holds String-related helper functions used only in tests.` might be better?

File: src/test/java/teammates/test/driver/TestUrl.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+public final class TestUrl {
+
+    private TestUrl() {
+        // utility class
+    }
+    /**
+     * Gets the relative path of a full URL. Useful for http/https-based URLs.
+     * @throws MalformedURLException if the given {@code url} is malformed
+     */
+
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestUrl.java]: The blank line should be before the header comment, not after.

File: src/test/java/teammates/test/driver/TestUrl.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package teammates.test.driver;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import teammates.common.util.Url;
+
+/**
+ *TestUrl.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestUrl.java]: Missing space. Also, `Utility methods for Url class used only in tests.` might be better?

File: src/main/java/teammates/common/util/FieldValidator.java (hunk 1 above)
Comment (whipermr5) [inline src/main/java/teammates/common/util/FieldValidator.java]: If inheritance is used, it can be made `protected` instead.

File: src/test/java/teammates/test/driver/TestUrl.java (hunk 11 above)
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestUrl.java]: If extending from `AppUrl` class, it can be `Additional AppUrl utility methods used only in tests.`

File: src/main/java/teammates/common/util/FieldValidator.java (hunk 1 above)
Comment (sujeet14108) [inline src/main/java/teammates/common/util/FieldValidator.java]: @whipermr5  Now using Inheritance for FieldValidator  but i don't think there is need of inheritance  in other classes.
Also if using inheritance i.e` teammates.test.driver.StringHelper  extends teammates.common.util.StringHelper` then there are other problems like making util.StringHelper as abstract class.
Please see the new changes

File: src/test/java/teammates/test/cases/util/FieldValidatorTest.java [hunk 12]
```diff
@@ -28,7 +29,8 @@ public void testGetValidityInfoForSizeCappedNonEmptyString() {
         int typicalLength = 25;
 
         try {
-            validator.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName, typicalLength, null);
+            teammates.test.driver.FieldValidator.getValidityInfoForSizeCappedNonEmptyString(typicalFieldName,
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: Why not import `teammates.test.driver.FieldValidator` and remove the import for `teammates.common.util.FieldValidator`? Then all these lines don't have to be changed.

File: src/test/java/teammates/test/driver/FieldValidator.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+     * @param fieldName
+     *            A descriptive name of the field e.g., "student name", to be
+     *            used in the return value to make the explanation more
+     *            descriptive.
+     * @param value
+     *            The string to be checked.
+     * @return An explanation of why the {@code value} is not acceptable.
+     *         Returns an empty string "" if the {@code value} is acceptable.
+     */
+    public static String getValidityInfoForSizeCappedNonEmptyString(String fieldName, int maxLength, String value) {
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/FieldValidator.java]: This can be non-static to match the original. This would also fix the PMD violation.

File: src/test/java/teammates/test/driver/TestUrl.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package teammates.test.driver;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import teammates.common.util.Url;
+
+/**
+ * Utility methods for Url class used only in tests.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestUrl.java]: On second thoughts, `Holds additional methods for {@link Url} used only in tests.`

File: src/test/java/teammates/test/driver/TestTimeHelper.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
+package teammates.test.driver;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.TimeZone;
+import teammates.common.util.TimeHelper;
+
+/**
+ * A helper class to hold time-related functions (e.g., converting dates to strings etc.).
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestTimeHelper.java]: `Holds additional methods for {@link TimeHelper} used only in tests.`

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package teammates.test.driver;
+
+import teammates.common.util.StringHelper;
+
+/**
+ * Holds String-related helper functions used only in tests.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: On second thoughts, `Holds additional methods for {@link StringHelper} used only in tests.`

File: src/test/java/teammates/test/driver/FieldValidator.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
+package teammates.test.driver;
+
+import teammates.common.util.Assumption;
+import teammates.common.util.SanitizationHelper;
+
+/**
+ * Used to handle the data validation aspect for test.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/FieldValidator.java]: `Holds additional methods for {@link teammates.common.util.FieldValidator} used only in tests.`

File: src/test/java/teammates/test/driver/TestStringHelper.java (hunk 16 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: @whipermr5  Check now :)

Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: This one hasn't been fixed yet.

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+public final class TestStringHelper {
+
+    private TestStringHelper() {
+        // utility class
+    }
+
+    /**
+     *generate string of given length.
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: This needs improvement. Refer to the other doc comments in `StringHelper.java`.

Comment (sujeet14108) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: @whipermr5 Done :)

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+    private TestStringHelper() {
+        // utility class
+    }
+
+    /**
+     * Generates string of given length.
+     * @param length of which string to be generated
+     * @return generated Strings
+     */
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Apologies; I realise now the method name is self-documenting so just remove the whole doc comment, like in the original.

Comment (sujeet14108) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Ok :)

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 20]
```diff
@@ -0,0 +1,24 @@
+package teammates.test.driver;
+
+import teammates.common.util.StringHelper;
+
+/**
+ * Holds additional methods for {@link StringHelper} used only in tests.
+ */
+
```
Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Also remove this blank line.

File: src/test/java/teammates/test/driver/TestStringHelper.java (hunk 19 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: It needs to have comment as per lint's rule.

Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Oh I see 😅 Slight improvements suggested:
`Generates an arbitrary string of given length.`
`@param length of string to be be generated`
`@return generated string`

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 21]
```diff
@@ -0,0 +1,23 @@
+package teammates.test.driver;
+
+import teammates.common.util.StringHelper;
+
+/**
+ * Holds additional methods for {@link StringHelper} used only in tests.
+ */
+public final class TestStringHelper {
```
Comment (damithc) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: @whipermr5 do we need the `Test` prefix? We did not add such a prefix for the `teammates.test.driver.FieldValidator` class. It could be a bit confusing because we use use `Test` as a suffix for test classes. I'm ok to use the prefix but in that case we should add it for all such classes? What do you think?

File: src/test/java/teammates/test/driver/TestStringHelper.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
+package teammates.test.driver;
```
Comment (damithc) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Perhaps we can put these in a `teammates.test.driver.util` package instead?

File: src/test/java/teammates/test/driver/TestStringHelper.java (hunk 21 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Why i used test as prefix was to make it more self explanatory but since as we have not used it for other classes so we can drop the prifix.
So what should we do?

Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: I'm actually not fond of the prefix, but it seems necessary if we need to import both the production class and its non-production extension (otherwise only one can be imported due to name collision, and the other would have to be referenced using its package name + class name).

Is it permissible to change the production classes to non-`final` and make the `private` constructors `protected` instead? This would allow the non-production version to extend the production class; only the non-production version would have to be imported then.

In summary, the options:

1. No prefix; use package name + class name when both production and non-production versions are referenced in the same test
2. No prefix; allow original util classes to be extended; only import non-production version into tests
3. Prefix all non-production versions; both can be imported normally into tests. It can also be a suffix instead of a prefix, e.g. `StringHelperExtension`?

Comment (damithc) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Option 2: I'd rather not test a class through a sub class because it's not clear what the SUT is.

Option 3: Something like `StringHelperTestHelper`?  It's a bit long but the suffix accurately describes the purpose of the class.

Comment (sujeet14108) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: @damithc  @whipermr5  I  would also prefer the Option 3 so should we proceed with that?  And how suffix should be used ?  `StringHelperTestHelper` or `StringHelperExtension` which should be used  former look more descriptive and fine to me except somewhat long

Comment (whipermr5) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Okay option 3 then. To me `StringHelperTestHelper` sounds like it's a helper that's used only in `StringHelperTest`, when in fact it's used in many tests.

Comment (damithc) [inline src/test/java/teammates/test/driver/TestStringHelper.java]: Let's use `StringHelperExtension` then.

**Comment by sujeet14108** [comment]
@whipermr5 Please check :)

File: src/test/java/teammates/test/cases/util/FieldValidatorTest.java [hunk 23]
```diff
@@ -1,19 +1,19 @@
 package teammates.test.cases.util;
 
 // CHECKSTYLE.OFF:AvoidStarImport as we want to perform tests on everything from FieldValidator
-import static teammates.common.util.FieldValidator.*;
-// CHECKSTYLE.ON:AvoidStarImport
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: Remember to turn checkstyle back on.

File: src/test/java/teammates/test/cases/util/FieldValidatorTest.java [hunk 24]
```diff
@@ -1,19 +1,19 @@
...
 
 import org.testng.annotations.Test;
 
 import com.google.appengine.api.datastore.Text;
 
-import teammates.common.util.FieldValidator;
 import teammates.common.util.StringHelper;
-import teammates.common.util.TimeHelper;
 import teammates.test.cases.BaseTestCase;
+import teammates.test.driver.FieldValidator;
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: Now that @damithc has weighed in that we should not test through a subclass, use the `Extension` suffix here as well. The extension can be a util class (final + private constructor) and the method static.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: @whipermr5  The extension classes are already util class like in TimeHelperExtension

    public final class TimeHelperExtension {

    private TimeHelperExtension() {
        // utility class
    }
`

Comment (whipermr5) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: I'm referring to `FieldValidatorExtension` :)

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/util/FieldValidatorTest.java]: @whipermr5 Sorry my bad. Would change it now :)

**Comment by sujeet14108** [comment]
@whipermr5  Done :+1:
